### PR TITLE
Move semantics for `sf::Drawable` , `sf::RenderTarget`, `sf::RenderTexture` and `sf::VertexBuffer`

### PR DESCRIPTION
**Pull request by Griezn**
## Description

I implemented the move semantics for `sf::Drawable` , `sf::RenderTarget`, `sf::RenderTexture` and `sf::VertexBuffer`.

This PR is related to the issue #2483 

## Tasks

* [ ] Tested on Linux
* [x] Tested on Windows
* [ ] Tested on macOS
* [ ] Tested on iOS
* [ ] Tested on Android

## How to test this PR?

Describe how to best test these changes. Please provide a [minimal, complete and verifiable example](https://stackoverflow.com/help/mcve) if possible, you can use the follow template as a start:

```cpp
#include <SFML/Graphics.hpp>
#include <iostream>

class CustomDrawable : public sf::Drawable {
public:
    explicit CustomDrawable(sf::VertexArray vertices) : m_vertices(std::move(vertices)) {}

private:
    void draw(sf::RenderTarget& target,const sf::RenderStates& states) const override {
        target.draw(m_vertices, states);
    }

    sf::VertexArray m_vertices;
};

int main() {
    // Create a render window
    sf::RenderWindow window(sf::VideoMode({ 800, 600 }), "Move Semantics Example");

    // Create a RenderTexture
    sf::RenderTexture renderTexture;
    if (!renderTexture.create({400, 300})) {
        std::cerr << "Failed to create RenderTexture" << std::endl;
        return 1;
    }
    sf::RenderTexture movedRenderTexture(std::move(renderTexture));
    std::cout << "moved RenderTexture" << std::endl;

    // Create a VertexBuffer
    sf::VertexBuffer vertexBuffer(sf::VertexBuffer::Static);
    if (!vertexBuffer.create(4)) {
        std::cerr << "Failed to create VertexBuffer" << std::endl;
        return 1;
    }
    sf::VertexBuffer movedVertexBuffer(std::move(vertexBuffer));
    std::cout << "moved vertexBuffer" << std::endl;

    // Create a custom drawable object
    sf::VertexArray vertices(sf::PrimitiveType::Triangles, 4);
    vertices[0].position = sf::Vector2f(0, 0);
    vertices[1].position = sf::Vector2f(100, 0);
    vertices[2].position = sf::Vector2f(100, 100);
    vertices[3].position = sf::Vector2f(0, 100);
    CustomDrawable customDrawable(std::move(vertices));

    CustomDrawable moveConstructedDrawable(std::move(customDrawable));

    CustomDrawable moveAssignedDrawable = std::move(moveConstructedDrawable);

    // Main loop
    while (window.isOpen()) {
        sf::Event event;
        while (window.pollEvent(event)) {
            if (event.type == sf::Event::Closed)
                window.close();
        }

        // Clear the RenderTexture
        movedRenderTexture.clear(sf::Color::White);

        // Draw the moved custom drawable object to the RenderTexture
        movedRenderTexture.draw(moveAssignedDrawable);

        // Display the contents of the RenderTexture on the window
        sf::Sprite sprite(movedRenderTexture.getTexture());
        window.clear(sf::Color::Black);
        window.draw(sprite);
        window.display();
    }

    return 0;
}
```

### CODE REVIEW

File: test/Graphics/Drawable.test.cpp [hunk 1]
```diff
@@ -23,14 +23,14 @@ class DrawableTest : public sf::Drawable
     mutable int m_callCount{};
 };
 
-TEST_CASE("[Graphics] sf::Drawable", runDisplayTests())
+TEST_CASE("[Graphics] sf::Drawable")
```
Comment (eXpl0it3r) [inline test/Graphics/Drawable.test.cpp:26]: You need to keep the `runDisplayTests()` as not all CI builds support tests that require a display

File: test/Graphics/RenderTarget.test.cpp [hunk 2]
```diff
@@ -1,9 +1,52 @@
...
+        return true;
+    }
+
+    bool setActive(bool) override
+    {
+        return true;
+    }
+};
+
+TEST_CASE("[Graphics] sf::RenderTarget")
```
Comment (eXpl0it3r) [inline test/Graphics/RenderTarget.test.cpp:26]: Most likely needs `runDisplayTests()` as well

File: test/Graphics/RenderTexture.test.cpp [hunk 3]
```diff
@@ -1,8 +1,33 @@
...
 
+#include <catch2/catch_test_macros.hpp>
+
 #include <type_traits>
 
-static_assert(!std::is_copy_constructible_v<sf::RenderTexture>);
-static_assert(!std::is_copy_assignable_v<sf::RenderTexture>);
-static_assert(!std::is_nothrow_move_constructible_v<sf::RenderTexture>);
-static_assert(!std::is_nothrow_move_assignable_v<sf::RenderTexture>);
+TEST_CASE("[Graphics] sf::RenderTexture")
```
Comment (eXpl0it3r) [inline test/Graphics/RenderTexture.test.cpp:7]: Need to use `runDisplayTests()`

**Review by eXpl0it3r — CHANGES_REQUESTED** [review]
Thank you for tackling those open points! 🙂

**Comment by ChrisThrasher** [comment]
The big problem with trying to add default move operations in the past is that certain classes have had data for which the default move operations were incorrect. For example OpenGL will sometimes give us an integer which refers to some resources which we're obligated to clean up in the destructor. The default move operations result in that resource being freed twice.

Have you gone through every data member of all these classes to ensure that the default move operations will in fact do what we want? To be honest it feels too good to be true that we can merely use default operations in all these places and get exactly the results we want.

File: include/SFML/Graphics/Drawable.hpp [hunk 4]
```diff
@@ -49,6 +49,36 @@ class SFML_GRAPHICS_API Drawable
     ////////////////////////////////////////////////////////////
     virtual ~Drawable() = default;
 
+    ////////////////////////////////////////////////////////////
+    /// \brief Default constructor
+    ///
+    ////////////////////////////////////////////////////////////
+    Drawable() = default;
```
Comment (ChrisThrasher) [inline include/SFML/Graphics/Drawable.hpp:56]: I'm not sure why these changes are necessary. This is not a class you can construct so it's not copy nor move constructible. It is however copy and move assignable. I'm not sure what any of this code or the changes you made in Drawable.test.cpp are accomplishing. I think you can remove it all.

File: test/Graphics/VertexBuffer.test.cpp [hunk 5]
```diff
@@ -90,6 +90,30 @@ TEST_CASE("[Graphics] sf::VertexBuffer", "[.display]")
         }
     }
 
+    SECTION("Move semantics")
+    {
+        SECTION("Construction")
+        {
+            sf::VertexBuffer       moveVertexBuffer(sf::PrimitiveType::LineStrip, sf::VertexBuffer::Dynamic);
```
Comment (ChrisThrasher) [inline test/Graphics/VertexBuffer.test.cpp:97]:
```suggestion
            sf::VertexBuffer       movedVertexBuffer(sf::PrimitiveType::LineStrip, sf::VertexBuffer::Dynamic);
```
Use a `moved` prefix instead of `move` to better match what is used elsewhere in the tests. Other files have this same inconsistency.

File: test/Graphics/RenderTarget.test.cpp [hunk 6]
```diff
@@ -1,9 +1,52 @@
...
+        STATIC_CHECK(!std::is_constructible_v<sf::RenderTarget>);
+        STATIC_CHECK(!std::is_copy_constructible_v<sf::RenderTarget>);
+        STATIC_CHECK(!std::is_copy_assignable_v<sf::RenderTarget>);
+        STATIC_CHECK(!std::is_nothrow_move_constructible_v<sf::RenderTarget>);
+        STATIC_CHECK(std::is_nothrow_move_assignable_v<sf::RenderTarget>);
+    }
+
+    SECTION("Move semantics")
+    {
+        SECTION("Move constructor")
```
Comment (ChrisThrasher) [inline test/Graphics/RenderTarget.test.cpp:39]:
```suggestion
        SECTION("Construction")
```

**Comment by Griezn** [comment]
Good point, it probably is too good to be true. I will perform some more testing on my end.

File: src/SFML/Graphics/RenderTexture.cpp [hunk 7]
```diff
@@ -45,6 +45,14 @@ RenderTexture::RenderTexture() = default;
 RenderTexture::~RenderTexture() = default;
 
 
+////////////////////////////////////////////////////////////
+RenderTexture::RenderTexture(RenderTexture&& other) noexcept = default;
```
Comment (ChrisThrasher) [inline src/SFML/Graphics/RenderTexture.cpp:49]:
```suggestion
RenderTexture::RenderTexture(RenderTexture&&) noexcept = default;
```

You can omit the parameter name for these defaulted functions.

File: src/SFML/Graphics/RenderTexture.cpp [hunk 8]
```diff
@@ -45,6 +45,14 @@ RenderTexture::RenderTexture() = default;
...
+
+
+////////////////////////////////////////////////////////////
+RenderTexture& RenderTexture::operator=(RenderTexture&& other) noexcept = default;
```
Comment (ChrisThrasher) [inline src/SFML/Graphics/RenderTexture.cpp:53]:
```suggestion
RenderTexture& RenderTexture::operator=(RenderTexture&&) noexcept = default;
```

**Review by ChrisThrasher — CHANGES_REQUESTED** [review]
I think there's something here we can merge. We just need to be careful we're getting it right since it's easy to mess up move semantics in subtle ways.

File: include/SFML/Graphics/Drawable.hpp (hunk 4 above)
Comment (Griezn) [reply include/SFML/Graphics/Drawable.hpp:56]: Because of the creation of the move constructor, the compiler saw the other constructors as implicitly deleted and thus causing build errors, if I recall correctly.

Comment (ChrisThrasher) [reply include/SFML/Graphics/Drawable.hpp:56]: It doesn't matter that you added `= default` to those constructors. This class contains a pure virtual function so it can never been constructible. You can tell this is true because the type traits for `sf::Drawable` did not change after you added all these defaulted functions. The class's functionality has not changed.

I checked out this PR and reverted all the changes to `sf::Drawable` and it built just fine for me.

File: include/SFML/Graphics/VertexBuffer.hpp
Comment (ChrisThrasher) [inline include/SFML/Graphics/VertexBuffer.hpp:1]: I don't think this is going to work. See how `VertexBuffer::~VertexBuffer()` frees `m_buffer`? If we use default move semantics then that will result in the resource being freed twice. Please revert all changes to `sf::VertexBuffer`.

File: include/SFML/Graphics/Drawable.hpp (hunk 4 above)
Comment (Griezn) [reply include/SFML/Graphics/Drawable.hpp:56]: I just tried again to be sure and this is what I get:
```
FAILED: src/SFML/Graphics/CMakeFiles/sfml-graphics.dir/VertexArray.cpp.obj 
C:\PROGRA~1\MICROS~2\2022\COMMUN~1\VC\Tools\MSVC\1436~1.325\bin\Hostx86\x86\cl.exe  /nologo /TP -DSFML_GRAPHICS_EXPORTS -DSTBI_FAILURE_USERMSG -D_CRT_SECURE_NO_WARNINGS -D_WINSOCK_DEPRECATED_NO_WARNINGS -IC:\Users\seppe\CLionProjects\SFML_fork\include -IC:\Users\seppe\CLionProjects\SFML_fork\src -external:IC:\Users\seppe\CLionProjects\SFML_fork\extlibs\headers\stb_image -external:IC:\Users\seppe\CLionProjects\SFML_fork\extlibs\headers\glad\include -external:IC:\Users\seppe\CLionProjects\SFML_fork\extlibs\headers\freetype2 -external:W0 /DWIN32 /D_WINDOWS /EHsc /Zi /Ob0 /Od /RTC1 -MDd /WX /W4 /w14242 /w14254 /w14263 /w14265 /w14287 /we4289 /w14296 /w14311 /w14545 /w14546 /w14547 /w14549 /w14555 /w14619 /w14640 /w14826 /w14905 /w14906 /w14928 /permissive- /wd4068 /wd4505 /wd4800 -std:c++17 /showIncludes /Fosrc\SFML\Graphics\CMakeFiles\sfml-graphics.dir\VertexArray.cpp.obj /Fdsrc\SFML\Graphics\CMakeFiles\sfml-graphics.dir\ /FS -c C:\Users\seppe\CLionProjects\SFML_fork\src\SFML\Graphics\VertexArray.cpp
C:\Users\seppe\CLionProjects\SFML_fork\src\SFML\Graphics\VertexArray.cpp(35): error C2512: 'sf::Drawable': no appropriate default constructor available
C:\Users\seppe\CLionProjects\SFML_fork\include\SFML/Graphics/Drawable.hpp(43): note: see declaration of 'sf::Drawable'
C:\Users\seppe\CLionProjects\SFML_fork\src\SFML\Graphics\VertexArray.cpp(35): note: see reference to function 'sf::VertexArray::VertexArray(void)'
C:\Users\seppe\CLionProjects\SFML_fork\src\SFML\Graphics\VertexArray.cpp(40): error C2512: 'sf::Drawable': no appropriate default constructor available
C:\Users\seppe\CLionProjects\SFML_fork\include\SFML/Graphics/Drawable.hpp(43): note: see declaration of 'sf::Drawable'
[64/105] Building CXX object src\SFML\Graphics\CMakeFiles\sfml-graphics.dir\ConvexShape.cpp.obj
[65/105] Building CXX object src\SFML\Graphics\CMakeFiles\sfml-graphics.dir\RenderTarget.cpp.obj
[66/105] Building CXX object src\SFML\Graphics\CMakeFiles\sfml-graphics.dir\RenderWindow.cpp.obj
[67/105] Building CXX object src\SFML\Graphics\CMakeFiles\sfml-graphics.dir\Sprite.cpp.obj
FAILED: src/SFML/Graphics/CMakeFiles/sfml-graphics.dir/Sprite.cpp.obj 
C:\PROGRA~1\MICROS~2\2022\COMMUN~1\VC\Tools\MSVC\1436~1.325\bin\Hostx86\x86\cl.exe  /nologo /TP -DSFML_GRAPHICS_EXPORTS -DSTBI_FAILURE_USERMSG -D_CRT_SECURE_NO_WARNINGS -D_WINSOCK_DEPRECATED_NO_WARNINGS -IC:\Users\seppe\CLionProjects\SFML_fork\include -IC:\Users\seppe\CLionProjects\SFML_fork\src -external:IC:\Users\seppe\CLionProjects\SFML_fork\extlibs\headers\stb_image -external:IC:\Users\seppe\CLionProjects\SFML_fork\extlibs\headers\glad\include -external:IC:\Users\seppe\CLionProjects\SFML_fork\extlibs\headers\freetype2 -external:W0 /DWIN32 /D_WINDOWS /EHsc /Zi /Ob0 /Od /RTC1 -MDd /WX /W4 /w14242 /w14254 /w14263 /w14265 /w14287 /we4289 /w14296 /w14311 /w14545 /w14546 /w14547 /w14549 /w14555 /w14619 /w14640 /w14826 /w14905 /w14906 /w14928 /permissive- /wd4068 /wd4505 /wd4800 -std:c++17 /showIncludes /Fosrc\SFML\Graphics\CMakeFiles\sfml-graphics.dir\Sprite.cpp.obj /Fdsrc\SFML\Graphics\CMakeFiles\sfml-graphics.dir\ /FS -c C:\Users\seppe\CLionProjects\SFML_fork\src\SFML\Graphics\Sprite.cpp
C:\Users\seppe\CLionProjects\SFML_fork\src\SFML\Graphics\Sprite.cpp(40): error C2512: 'sf::Drawable': no appropriate default constructor available
C:\Users\seppe\CLionProjects\SFML_fork\include\SFML/Graphics/Drawable.hpp(43): note: see declaration of 'sf::Drawable'
C:\Users\seppe\CLionProjects\SFML_fork\src\SFML\Graphics\Sprite.cpp(47): error C2512: 'sf::Drawable': no appropriate default constructor available
C:\Users\seppe\CLionProjects\SFML_fork\include\SFML/Graphics/Drawable.hpp(43): note: see declaration of 'sf::Drawable'
[68/105] Building CXX object src\SFML\Graphics\CMakeFiles\sfml-graphics.dir\Shader.cpp.obj
ninja: build stopped: subcommand failed.
```

File: include/SFML/Graphics/VertexBuffer.hpp
Comment (Griezn) [reply include/SFML/Graphics/VertexBuffer.hpp:1]: I'll have a look at that tomorrow!

File: include/SFML/Graphics/Drawable.hpp (hunk 4 above)
Comment (ChrisThrasher) [reply include/SFML/Graphics/Drawable.hpp:56]: https://github.com/ChrisThrasher/SFML/actions/runs/5193825019/jobs/9364846312

I removed the changes to `sf::Drawable` and the CI pipeline is happy with it. I think something is weird about your local build environment but I'm not sure what.

**Comment by ChrisThrasher** [comment]
https://github.com/ChrisThrasher/SFML/commit/4a788711010196621d676ac8c6045ba53d33229c

I made some simplifications. We don't write tests for default move semantics so I simplified the changes to the tests. I removed the changes to `VertexBuffer` since we know default move semantics are incorrect there. The PR becomes a lot more simple. It's just adding a few default operations to `sf::RenderTexture` and `sf::RenderTarget`. Those classes do not have move semantics enabled because they both have destructors. Even though the destructors are defaulted, that still implicitly disables move semantics. There no other reason (as far as I'm aware) that we can't move those types.

File: test/Graphics/RenderTarget.test.cpp [hunk 9]
```diff
@@ -1,9 +1,53 @@
...
+        }
+
+        SECTION("Assignment")
+        {
+            RenderTargetTest moveRenderTarget;
+            RenderTargetTest renderTarget;
+            renderTarget = std::move(moveRenderTarget);
+        }
+    }
+}
```
Comment (kimci86) [inline test/Graphics/RenderTarget.test.cpp:53]: We usually end files with a blank line. It is missing here.

File: test/Graphics/RenderTexture.test.cpp [hunk 10]
```diff
@@ -1,8 +1,34 @@
...
+        }
+
+        SECTION("Move assignment")
+        {
+            sf::RenderTexture moveRenderTexture;
+            sf::RenderTexture renderTexture;
+            renderTexture = std::move(moveRenderTexture);
+        }
+    }
+}
```
Comment (kimci86) [inline test/Graphics/RenderTexture.test.cpp:34]: Same here, missing a final blank line.

**Review by kimci86 — COMMENTED** [review]
Besides the feedback already given, my only remaining concern is if `RenderTarget::m_id` can safely be moved from one instance to another especially between different threads.

**Comment by ChrisThrasher** [comment]
The precedent we have set is to not write tests for default move semantics so you can remove all the tests you added. We only manually test move semantics when we wrote them by hand.
